src/ProtectedRoomsConfig.ts

```diff
@@ -0,0 +1,127 @@
+/*
+Copyright 2019, 2022 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import AwaitLock from 'await-lock';
+import { extractRequestError, LogService, MatrixClient, Permalinks } from "matrix-bot-sdk";
+import { IConfig } from "./config";
+const PROTECTED_ROOMS_EVENT_TYPE = "org.matrix.mjolnir.protected_rooms";
+
+/**
+ * Manages the set of rooms that the user has explicitly asked to be protected.
+ */
+export default class ProtectedRoomsConfig {
+
+    /**
+     * These are rooms that we explictly asked Mjolnir to protect, usually via the `rooms add` command.
+     * They are not all of the rooms that mjolnir is protecting as with `config.protectAllJoinedRooms`.
+     */
+    private explicitlyProtectedRooms = new Set</*room id*/string>();
+    /** This is to prevent clobbering the account data for the protected rooms if several rooms are explicitly protected concurrently. */
+    private accountDataLock = new AwaitLock();
+
+    constructor(private readonly client: MatrixClient) {
+
+    }
+
+    /**
+     * Load any rooms that have been explicitly protected from a Mjolnir config.
+     * Will also ensure we are able to join all of the rooms.
+     * @param config The config to load the rooms from under `config.protectedRooms`.
+     */
+    public async loadProtectedRoomsFromConfig(config: IConfig): Promise<void> {
+        // Ensure we're also joined to the rooms we're protecting
+        LogService.info("ProtectedRoomsConfig", "Resolving protected rooms...");
+        const joinedRooms = await this.client.getJoinedRooms();
+        for (const roomRef of config.protectedRooms) {
+            const permalink = Permalinks.parseUrl(roomRef);
+            if (!permalink.roomIdOrAlias) continue;
+
+            let roomId = await this.client.resolveRoom(permalink.roomIdOrAlias);
+            if (!joinedRooms.includes(roomId)) {
+                roomId = await this.client.joinRoom(permalink.roomIdOrAlias, permalink.viaServers);
```

Comment:
If I read this correctly, we stop the entire initialization if we can't join one of these rooms, right?

Is that by design?

Reply:
Yep that's how it reads. That's how it was before, i think I agree with this and that it was probably intentional, given that the config file isn't something that can somehow be corrupted overtime through operation and if it does, then the operator needs to know about it.